src/tactic/induction.lean

```diff
@@ -0,0 +1,1290 @@
+  state ← read,
+  trace_eliminate_hyp $ format.join
+    [to_fmt (msg ()), "\n-----\n", to_fmt state, "\n-----"]
+
+
+/-!
+## Information Gathering
+
+We define data structures for information relevant to the induction, and
+functions to collect this information for a specific goal.
+-/
+
+
+/--
+Information about a constructor argument. Contains:
+
+- `aname`: the argument's name.
+- `type` : the argument's type.
+- `dependent`: whether the argument is dependent, i.e. whether it occurs in the
+  remainder of the constructor type.
+- `index_occurrences`: the index arguments of the constructor's return type
+  in which this argument occurs. If the constructor return type is
+  `I i₀ ... iₙ` and the argument under consideration is `a`, and `a` occurs in
+  `i₁` and `i₂`, then the `index_occurrences` are `1, 2`. As an additional
+  requirement, for `iⱼ` to be considered an index occurrences,
+  the type of `iⱼ` must match that of `a` according to
+  `index_occurrence_type_match`.
+- `is_recursive`: whether this is a recursive constructor argument.
+-/
+@[derive has_reflect]
+meta structure constructor_argument_info :=
+(aname : name)
+(type : expr)
+(dependent : bool)
+(index_occurrences : list ℕ)
+(is_recursive : bool)
+
+/--
+Information about a constructor. Contains:
+
+- `cname`: the constructor's name.
+- `non_param_args`: information about the arguments of the constructor,
+  excluding the arguments induced by the parameters of the inductive type.
+- `num_non_param_args`: the length of `non_param_args`.
+- `rec_args`: the subset of `non_param_args` which are recursive constructor
+  arguments.
+- `num_rec_args`: the length of `rec_args`.
+
+For example, take the constructor
+```
+list.cons : ∀ {α} (x : α) (xs : list α), list α
+```
+`α` is a parameter of `list`, so `non_param_args` contains information about `x`
+and `xs`. `rec_args` contains information about `xs`.
+-/
+@[derive has_reflect]
+meta structure constructor_info :=
+(cname : name)
+(non_param_args : list constructor_argument_info)
+(num_non_param_args : ℕ)
+(rec_args : list constructor_argument_info)
+(num_rec_args : ℕ)
+
+/--
+When we construct the goal for the minor premise of a given constructor, this is
+the number of hypotheses we must name.
+-/
+meta def constructor_info.num_nameable_hypotheses (c : constructor_info) : ℕ :=
+c.num_non_param_args + c.num_rec_args
+
+/--
+Information about an inductive type. Contains:
+
+- `iname`: the type's name.
+- `constructors`: information about the type's constructors.
+- `num_constructors`: the length of `constructors`.
+- `type`: the type's type.
+- `num_param`: the type's number of parameters.
+- `num_indices`: the type's number of indices.
+-/
+@[derive has_reflect]
+meta structure inductive_info :=
+(iname : name)
+(constructors : list constructor_info)
+(num_constructors : ℕ)
+(type : expr)
+(num_params : ℕ)
+(num_indices : ℕ)
+
+/--
+Information about an eliminee (i.e. the hypothesis on which we are performing
```

Comment:
This is usually called the `major premise`.  Is the name *eliminee* used for this hypothesis? I've never heard it.

Reply:
"Eliminee" is my invention. I don't recall why I wanted this instead of "major premise". Maybe because it's shorter? Anyway, I should probably change it to follow convention.